### PR TITLE
Tesla Model 3 / Model Y Car Port (Clean)

### PR DESCRIPTION
**Pull request by lukasloetkolben**
In a conversation with DebboR, we considered removing the Model S code from openpilot and adding a clean Model3 code instead. Here is an implementation with only Tesla Model 3/Y.

The best thing about this carport is that only a single bus had to be intercepted. 🎉
No external Panda is required, just a C3X and a harness box.

**Checklist Model 3**
- [x] added entry to CAR in selfdrive/car/tesla/values.py
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with openpilot: b5f2b5b4ef23f4b6/00000008--d88c2210a5
- [x] route with stock system: b5f2b5b4ef23f4b6/00000009--f3a74d938c
- [x] car harness used (if comma doesn't sell it, put N/A): N/A

**Checklist Model Y**
- [x] added entry to CAR in selfdrive/car/tesla/values.py
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with openpilot: 35f334fe58c5cdc5/00000089--c5ed862a40
- [x] route with stock system: 35f334fe58c5cdc5/00000001--d8d176b1af
- [x] car harness used (if comma doesn't sell it, put N/A): N/A

Tesla is in dashcam mode only, so Model 3 and Model Y will not yet appear in CARS.md or release notes.

**Additional Routes for the other Fingerprints:**
Model 3 2023 (RHD): 574d45dbc63e9fd8/00000000--0c80908e8c
Model 3 2023 (LHD): 8b67b428fd9fea5c/00000047--d04a4d7f8a
Model 3 2021 (LHD): 5df538ba7679f5b7/00000002--ab29e12a18
Model 3 2020 (LHD): be7997c2a9399899/00000001--9d87515889
Model Y 2022 (LHD): 58a1cdc98298efcb/00000000--166a842612
Model Y 2022 (RHD): 5313f747976147e6/00000015--c1fa5e2ce0
Model Y 2020 (LHD): 7588270b74f42d15/00000000--85ec85024a

**Demos on Youtube:**
https://www.youtube.com/watch?v=CwjCp1Gb9d4
https://www.youtube.com/watch?v=1rh0sfZPymE
https://www.youtube.com/watch?v=LC0k8hmkBus
https://www.youtube.com/watch?v=ljMDHocoxxM

### CODE REVIEW

**Comment by mattwebbio** [comment]
Is there anything preventing this from being merged?

**Comment by adeebshihadeh** [comment]
> Is there anything preventing this from being merged?

This is a substantially new car port, which takes time to review.

File: selfdrive/car/tesla/carcontroller.py [hunk 1]
```diff
@@ -40,23 +41,24 @@ def update(self, CC, CS, now_nanos):
...
-    if self.frame % 10 == 0 and pcm_cancel_cmd:
-      # Spam every possible counter value, otherwise it might not be accepted
-      for counter in range(16):
-        can_sends.append(self.tesla_can.create_action_request(CS.msg_stw_actn_req, pcm_cancel_cmd, CANBUS.chassis, counter))
-        can_sends.append(self.tesla_can.create_action_request(CS.msg_stw_actn_req, pcm_cancel_cmd, CANBUS.autopilot_chassis, counter))
+    # Sent cancel request only if ACC is enabled
+    if self.frame % 10 == 0 and pcm_cancel_cmd and CS.acc_enabled:
+      counter = int(CS.sccm_right_stalk["SCCM_rightStalkCounter"] + 1) % 16
+      can_sends.append(self.tesla_can.right_stalk_press(counter, 1))  # half up (cancel acc)
+      can_sends.append(self.tesla_can.right_stalk_press((counter + 1) % 16, 0))  # to prevent neutral gear warning
```
Comment (robbederks) [inline selfdrive/car/tesla/carcontroller.py:61]: little weird that it's `% 16` here again, let's make the initial counter variable be the actual current counter, and then `+1` and `+2` with the `% 16` on each line

File: selfdrive/car/tesla/carstate.py [hunk 2]
```diff
@@ -29,78 +28,80 @@ def update(self, cp, cp_cam):
...
     ret.brake = 0
-    ret.brakePressed = bool(cp.vl["BrakeMessage"]["driverBrakeStatus"] != 1)
+    ret.brakePressed = cp.vl["IBST_status"]["IBST_driverBrakeApply"] == 2
 
     # Steering wheel
-    epas_status = cp_cam.vl["EPAS3P_sysStatus"] if self.CP.carFingerprint == CAR.TESLA_MODELS_RAVEN else cp.vl["EPAS_sysStatus"]
+    epas_status = cp.vl["EPAS3S_sysStatus"]
+    self.hands_on_level = epas_status["EPAS3S_handsOnLevel"]
+    self.steer_warning = self.can_define.dv["EPAS3S_sysStatus"]["EPAS3S_eacErrorCode"].get(int(epas_status["EPAS3S_eacErrorCode"]), None)
+    steer_status = self.can_define.dv["EPAS3S_sysStatus"]["EPAS3S_eacStatus"].get(int(epas_status["EPAS3S_eacStatus"]), None)
```
Comment (robbederks) [inline selfdrive/car/tesla/carstate.py:43]: seems to be only used once, let's maybe inline

File: selfdrive/car/tesla/carstate.py [hunk 3]
```diff
@@ -29,78 +28,80 @@ def update(self, cp, cp_cam):
...
         event.pressed = state
-        buttonEvents.append(event)
+        button_events.append(event)
       self.button_states[button.event_type] = state
-    ret.buttonEvents = buttonEvents
+    ret.buttonEvents = button_events
 
     # Doors
-    ret.doorOpen = any((self.can_define.dv["GTW_carState"][door].get(int(cp.vl["GTW_carState"][door]), "OPEN") == "OPEN") for door in DOORS)
+    ret.doorOpen = any([cp_adas.vl["VCLEFT_doorStatus"]["VCLEFT_frontLatchSwitch"] != 1,
```
Comment (robbederks) [inline selfdrive/car/tesla/carstate.py:81]: let's clean that up with list of tuples like `("VCLEFT_doorStatus", "VCLEFT_frontLatchSwitch", 1)`

Comment (robbederks) [reply selfdrive/car/tesla/carstate.py:81]: maybe let's append the enums to the DBC like the model S port had

File: selfdrive/car/tesla/carstate.py [hunk 4]
```diff
@@ -29,78 +28,80 @@ def update(self, cp, cp_cam):
...
 
     # AEB
     ret.stockAeb = (cp_cam.vl["DAS_control"]["DAS_aebEvent"] == 1)
 
     # Messages needed by carcontroller
-    self.msg_stw_actn_req = copy.copy(cp.vl["STW_ACTN_RQ"])
-    self.acc_state = cp_cam.vl["DAS_control"]["DAS_accState"]
-    self.das_control_counters.extend(cp_cam.vl_all["DAS_control"]["DAS_controlCounter"])
+    self.sccm_right_stalk = copy.copy(cp_adas.vl["SCCM_rightStalk"])
+    self.das_control = copy.copy(cp_cam.vl["DAS_control"])
```
Comment (robbederks) [inline selfdrive/car/tesla/carstate.py:104]: no need to copy the whole message, let's save just the counter val which is the only one used?

File: selfdrive/car/tesla/teslacan.py [hunk 5]
```diff
@@ -86,9 +41,24 @@ def create_longitudinal_commands(self, acc_state, speed, min_accel, max_accel, c
       "DAS_controlCounter": cnt,
       "DAS_controlChecksum": 0,
     }
+    data = self.packer.make_can_msg("DAS_control", CANBUS.party, values)[2]
+    values["DAS_controlChecksum"] = self.checksum(0x2b9, data[:7])
+    return self.packer.make_can_msg("DAS_control", CANBUS.party, values)
+
+  def right_stalk_press(self, counter, position):
+    # TODO: Implement CRC checksum instead of lookup table.
```
Comment (robbederks) [inline selfdrive/car/tesla/teslacan.py:49]: would be _really_ nice to fix this

Comment (lukasloetkolben) [reply selfdrive/car/tesla/teslacan.py:49]: I have been working on this for a long time. I have not been able to work out the calculation using any method. I have tried CRC-beagle, revng. I tried to calculate it manually, but without success.

Perhaps there is a clever mind out there who can solve this?

Comment (robbederks) [reply selfdrive/car/tesla/teslacan.py:49]: @gregjhogan interested in trying? can upload a binary into ghidra if you want :)

Comment (dzid26) [reply selfdrive/car/tesla/teslacan.py:49]: @lukasloetkolben  have you tried including message id before message bytes when doing reverse eng?

Comment (gregjhogan) [reply selfdrive/car/tesla/teslacan.py:49]: Here is the algorithm for`SCCM_rightStalk` tested on a few messages:
```python
import crcmod

msgs = [
  b'\x8c\x0b\x00',
  b'\x31\x0e\x00',
  b'\x44\x01\x00',
]

right_stalk_init_val = [0x7C, 0xB6, 0xF0, 0x2F, 0x69, 0xA3, 0xDD, 0x1C, 0x56, 0x90, 0xCA, 0x09, 0x43, 0x7D, 0xB7, 0xF1]

for msg in msgs:
  cntr = msg[1] & 0xF
  crc1_func = crcmod.mkCrcFun(0x12f, initCrc=0x00, xorOut=0xFF, rev=False)
  crc1 = crc1_func(msg[1:]) & 0xFF
  crc2_func = crcmod.mkCrcFun(0x12f, initCrc=crc1, xorOut=0xFF, rev=False)
  crc2 = crc2_func(bytes([right_stalk_init_val[cntr]])) & 0xFF
  print(f'message: 0x{msg.hex()} -> crc: {crc2:02X} = {"GOOD" if crc2 == msg[0] else "BAD"}')
```

Comment (lukasloetkolben) [reply selfdrive/car/tesla/teslacan.py:49]: just tested it on some more data. I will implement it later. thank you!

**Comment by jayvin100** [comment]
If i have the 2024 M3 Highland will this work? I assume just have to add the fingerprint once I get it plugged in right?

**Comment by lukasloetkolben** [comment]
> If i have the 2024 M3 Highland will this work? I assume just have to add the fingerprint once I get it plugged in right?

This is only for Model3 with HW3. The model3 Highland has HW4 at least has a different connector, I can't say anything about the software yet.

**Pull request closed without merging**